src/stream.rs

```diff
@@ -0,0 +1,263 @@
+        ReadDecoder {
+            decoder: self,
+            stream: stream,
+            reached_stream_eof: false,
+            buffer: buffer,
+            unused_buffer_slice: 0..0,
+        }
+    }
+
+    pub fn wrap_write<S, B>(self, stream: S, buffer: B) -> WriteDecoder<S, B>
+    where S: WriteUnicode, B: AsMut<[u8]> {
+        WriteDecoder {
+            decoder: self,
+            stream: stream,
+            buffer: buffer,
+        }
+    }
+}
+
+impl Encoder {
+    pub fn wrap_read<S, B>(self, stream: S, buffer: B) -> ReadEncoder<S, B>
+    where S: ReadUnicode, B: AsMut<str> {
+        ReadEncoder {
+            encoder: self,
+            stream: stream,
+            reached_stream_eof: false,
+            buffer: buffer,
+            unused_buffer_slice: 0..0,
+        }
+    }
+
+    pub fn wrap_write<S, B>(self, stream: S, buffer: B) -> WriteEncoder<S, B>
+    where S: WriteBytes, B: AsMut<[u8]> {
+        WriteEncoder {
+            encoder: self,
+            stream: stream,
+            buffer: buffer,
+        }
+    }
+}
+
+impl<'a, S: ReadUnicode> ReadUnicode for &'a mut S {
+    fn read(&mut self, dst: &mut str) -> io::Result<usize> {
+        (**self).read(dst)
+    }
+}
+
+/// Similar to `std::io::Read for &'a [u8]`.
+impl<'a> ReadUnicode for &'a str {
+    fn read(&mut self, dst: &mut str) -> io::Result<usize> {
+        // Unsafe: copy_maximum_utf8_prefix carefully preserves UTF-8 wel-formedness.
+        let dst_bytes = unsafe { mem::transmute::<&mut str, &mut [u8]>(dst) };
+        let (bytes_copied, remaining_self) = copy_maximum_utf8_prefix(*self, dst_bytes);
+        *self = remaining_self;
+        Ok(bytes_copied)
+    }
+}
+
+fn copy_maximum_utf8_prefix<'a>(src: &'a str, dst: &mut [u8]) -> (usize, &'a str) {
+    let src_bytes = src.as_bytes();
+    let bytes_to_copy = if dst.len() >= src.len() {
+        src.len()
+    } else {
+        let mut i = dst.len() - 1;
+        while (src_bytes[i] & 0xC0) == 0x80 {
+            i -= 1;
+        }
+        i
+    };
+    let (to_copy_from, remaining_src) = src.split_at(bytes_to_copy);
+    let (to_copy_into, remaining_dst) = dst.split_at_mut(bytes_to_copy);
+    to_copy_into.copy_from_slice(to_copy_from.as_bytes());
+    for byte in remaining_dst {
+        if (*byte & 0xC0) == 0x80 {
+            *byte = 0
+        } else {
+            break
+        }
+    }
+    (bytes_to_copy, remaining_src)
+}
+
+
+// FIXME: with vs without replacement
+
+pub struct ReadDecoder<Stream: ReadBytes, Buffer: AsMut<[u8]>> {
+    decoder: Decoder,
+    stream: Stream,
+    reached_stream_eof: bool,
+    buffer: Buffer,
+    unused_buffer_slice: Range<usize>,
+}
+
+impl<Stream: ReadBytes, Buffer: AsMut<[u8]>> ReadUnicode for ReadDecoder<Stream, Buffer> {
+    fn read(&mut self, dst: &mut str) -> io::Result<usize> {
+        // Unsafe: Like Decoder::decode_to_utf8, <ReadDecoder as ReadBytes>::read’s contract
+        // is to preserve UTF-8 well-formedness.
```

Comment:
The comment is incorrect. `Decoder::decode_to_utf8` is documented (in the doc for the struct) _not_ to preserve the UTF-8 well-formedness of the tail of the slice (it may even write garbage in the tail as opposed to just leaving it the way it was). Only the `_to_str` version preserves the UTF-8 validity of the tail of the slice.